### PR TITLE
vlek: Adding vlek ioctls

### PR DESCRIPTION
**Pull request by larrydewey**
Adding in ioctls for SNP_VLEK_LOAD to allow platform owners to load VLEK where appropriate.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -238,6 +249,35 @@ impl std::convert::From<CertError> for UserApiError {
...
+    /// Certificate length does not match what was specified in the buffer.
+    InvalidLength,
+
+    /// No certificates were provided
+    EmptyCertBuffer,
+
+    /// Unknown Error.
+    UnknownError,
+}
+
```
Comment (DGonzalezVillal) [inline src/error.rs:264]: Why create a VLEK error instead of using the existing Cert error?

File: src/firmware/linux/host/types/snp.rs [hunk 2]
```diff
@@ -249,6 +249,71 @@ impl Default for SnpSetConfig {
...
+
+#[cfg(feature = "snp")]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[repr(C, packed)]
+/// Wrapped VLEK data.
+pub struct WrappedVlekHashstick<'a> {
+    /// Opaque data provided by AMD Key Distribution Server
+    /// (as described in SEV-SNP Firmware ABI 1.54, SNP_VLEK_LOAD)
+    pub data: &'a [u8], // 432 bytes of data
+}
```
Comment (DGonzalezVillal) [inline src/firmware/linux/host/types/snp.rs:263]: So the KDS provides bytes not an actual x509 cert?

File: src/firmware/linux/host/types/snp.rs [hunk 3]
```diff
@@ -249,6 +249,71 @@ impl Default for SnpSetConfig {
...
+
+#[cfg(feature = "snp")]
+impl SnpVlekLoad {
+    /// Creates a new VLEK load instruction from a hashstick.
+    pub fn new(hashstick: &WrappedVlekHashstick) -> Self {
+        hashstick.into()
+    }
+}
+
```
Comment (DGonzalezVillal) [inline src/firmware/linux/host/types/snp.rs:305]: So the use scenario for a user is that they provide the raw bytes into a SNPVlekLoad struct?

**Review by DGonzalezVillal — COMMENTED** [review]
Just a couple of discussion points, but the code looks good to me.

File: src/error.rs (hunk 1 above)
Comment (larrydewey) [reply src/error.rs:264]: When the platform owner calls SNP_VLEK_LOAD, they are passing a Wrapped Hashstick, not the x509 certificate itself. The certificate is provided to the guest via the extended guest request workflow.

Comment (larrydewey) [reply src/error.rs:264]: These errors are separate, as they are not certificate errors per se.

File: src/firmware/linux/host/types/snp.rs (hunk 2 above)
Comment (larrydewey) [reply src/firmware/linux/host/types/snp.rs:263]: Correct. It is a wrapped hashstick

File: src/firmware/linux/host/types/snp.rs (hunk 3 above)
Comment (larrydewey) [reply src/firmware/linux/host/types/snp.rs:305]: In this case, the user will provide the hashstick blob to the API, and the API will wrap the contents into a `WrappedVlekHashstick`, then that will be wrapped into the `SnpVlekLoad` command structure, which is subsequently passed to the IOCTL. Once the hashstick is loaded into the ASP, the platform owner will provide the certificate to the guest via the extended guest request mechanism.

File: src/error.rs (hunk 1 above)
Comment (larrydewey) [reply src/error.rs:264]: Do you think maybe we refer to them has `HashstickErrors`, instead?

**Review by DGonzalezVillal — APPROVED** [review]
LGTM

**Pull request merged**